File: email-proxy.js
// Simple email proxy server using Express.js
// This should be deployed to a server or cloud function

// Load environment variables
require('dotenv').config({ path: '.env.proxy' });

const express = require('express');
const cors = require('cors');
const { Resend } = require('resend');
const nodemailer = require('nodemailer');

const app = express();
app.use(cors());
app.use(express.json());

// Check if API key is provided
console.log('Resend API Key configured:', !!process.env.RESEND_API_KEY);
console.log('API Key starts with "re_"?:', process.env.RESEND_API_KEY && process.env.RESEND_API_KEY.startsWith('re_'));

// Initialize Resend with your API key
// In production, use environment variables
const resend = new Resend(process.env.RESEND_API_KEY);

// Create SMTP transporter if SMTP credentials are provided
let smtpTransporter = null;
if (process.env.EMAILJS_SMTP_HOST && process.env.EMAILJS_SMTP_USER && process.env.EMAILJS_SMTP_PASS) {
  smtpTransporter = nodemailer.createTransport({
    host: process.env.EMAILJS_SMTP_HOST,
    port: parseInt(process.env.EMAILJS_SMTP_PORT || '587', 10),
    secure: false, // true for 465, false for other ports
    auth: {
      user: process.env.EMAILJS_SMTP_USER,
      pass: process.env.EMAILJS_SMTP_PASS,
    },
    tls: {
      rejectUnauthorized: false // Set to true in production with proper certificates
    }
  });

  // Verify SMTP connection
  smtpTransporter.verify((error, success) => {
    if (error) {
      console.error('SMTP connection error:', error);
    } else {
      console.log('SMTP server is ready to take messages');
    }
  });
}

app.options('/send-email', (req, res) => {
  res.header('Access-Control-Allow-Origin', '*');
  res.header('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.header('Access-Control-Allow-Headers', 'Content-Type');
  res.sendStatus(200);
});

app.post('/send-email', async (req, res) => {
  try {
    console.log('Received email request:', req.body);
    
    const { from, to, subject, html } = req.body;
    
    // Validate input
    if (!to || !subject || !html) {
      return res.status(400).json({ 
        error: 'Missing required fields: to, subject, html' 
      });
    }
    
    // Check if Resend is properly configured
    if (!process.env.RESEND_API_KEY) {
      return res.status(500).json({ 
        error: 'Resend API key not configured',
        details: 'Please set RESEND_API_KEY in .env.proxy file'
      });
    }
    
    // Send email using Resend
    const data = await resend.emails.send({
      from: from || process.env.FROM_EMAIL || 'onboarding@resend.dev',
      to: Array.isArray(to) ? to : [to],
      subject,
      html,
    });
    
    console.log('Email sent successfully:', data);
    res.json({ 
      success: true, 
      data 
    });
  } catch (error) {
    console.error('Email sending error:', error);
    
    // Handle specific error cases
    if (error.message && error.message.includes('Unauthorized')) {
      return res.status(401).json({ 
        error: 'Invalid Resend API key',
        details: 'Please check your RESEND_API_KEY environment variable'
      });
    }
    
    if (error.message && error.message.includes('rate_limit')) {
      return res.status(429).json({ 
        error: 'Rate limit exceeded',
        details: 'Too many requests, please try again later'
      });
    }
    
    res.status(500).json({ 
      error: 'Failed to send email',
      details: error.message 
    });
  }
});

// SMTP bulk email endpoint
app.options('/smtp-bulk-email', (req, res) => {
  res.header('Access-Control-Allow-Origin', '*');
  res.header('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.header('Access-Control-Allow-Headers', 'Content-Type');
  res.sendStatus(200);
});

app.post('/smtp-bulk-email', async (req, res) => {
  try {
    console.log('Received SMTP bulk email request:', req.body);
    
    const { recipients, subject, html, text } = req.body;
    
    // Validate input
    if (!recipients || !Array.isArray(recipients) || recipients.length === 0) {
      return res.status(400).json({ 
        error: 'Missing or invalid recipients array' 
      });
    }
    
    if (!subject || !html) {
      return res.status(400).json({ 
        error: 'Missing required fields: subject, html' 
      });
    }
    
    // Check if SMTP is configured
    if (!smtpTransporter) {
      return res.status(500).json({ 
        error: 'SMTP not configured',
        details: 'Please set EMAILJS_SMTP_HOST, EMAILJS_SMTP_USER, and EMAILJS_SMTP_PASS in .env.proxy file'
      });
    }
    
    let sentCount = 0;
    let failedCount = 0;
    const errors = [];
    
    // Send emails to all recipients
    for (const recipient of recipients) {
      try {
        const mailOptions = {
          from: process.env.FROM_EMAIL || 'no-reply@suprisesuper.com',
          to: recipient.email,
          subject: subject,
          html: html,
          text: text,
        };
        
        await smtpTransporter.sendMail(mailOptions);
        sentCount++;
        console.log(`Email sent successfully to ${recipient.email}`);
      } catch (error) {
        failedCount++;
        errors.push(`Failed to send to ${recipient.email}: ${error.message}`);
        console.error(`Error sending email to ${recipient.email}:`, error);
      }
      
      // Add delay to respect rate limits
      await new Promise(resolve => setTimeout(resolve, 50));
    }
    
    res.json({ 
      success: true,
      sent: sentCount,
      failed: failedCount,
      errors: errors
    });
  } catch (error) {
    console.error('SMTP bulk email error:', error);
    res.status(500).json({ 
      error: 'Failed to send bulk emails via SMTP',
      details: error.message 
    });
  }
});

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({ 
    status: 'ok', 
    timestamp: new Date().toISOString(),
    resendApiKeyConfigured: !!process.env.RESEND_API_KEY,
    resendApiKeyValidFormat: process.env.RESEND_API_KEY && process.env.RESEND_API_KEY.startsWith('re_'),
    smtpConfigured: !!smtpTransporter
  });
});

const PORT = process.env.PORT || 3001;
app.listen(PORT, () => {
  console.log(`Email proxy server running on port ${PORT}`);
  console.log(`Health check: http://localhost:${PORT}/health`);
  console.log(`Send email endpoint: http://localhost:${PORT}/send-email`);
  console.log(`SMTP bulk email endpoint: http://localhost:${PORT}/smtp-bulk-email`);
  
  // Check if API key is configured
  if (!process.env.RESEND_API_KEY) {
    console.warn('WARNING: RESEND_API_KEY is not set in .env.proxy file');
    console.warn('Email sending will not work until you configure the API key');
  } else if (!process.env.RESEND_API_KEY.startsWith('re_')) {
    console.warn('WARNING: RESEND_API_KEY does not start with "re_" - it may be invalid');
  } else {
    console.log('Resend API key is configured correctly');
  }
  
  // Check if SMTP is configured
  if (smtpTransporter) {
    console.log('SMTP is configured and ready');
  } else {
    console.warn('SMTP is not configured - bulk emails will fall back to Resend API');
  }
});

File: email-proxy.py
# Simple email proxy server using Python Flask
# This is an alternative to the Node.js version

from flask import Flask, request, jsonify
from flask_cors import CORS
import requests
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv('.env.proxy')

app = Flask(__name__)
CORS(app)

RESEND_API_KEY = os.getenv('RESEND_API_KEY')
FROM_EMAIL = os.getenv('FROM_EMAIL', 'onboarding@resend.dev')

@app.route('/health', methods=['GET'])
def health():
    return jsonify({
        'status': 'ok',
        'timestamp': __import__('datetime').datetime.utcnow().isoformat(),
        'resendApiKeyConfigured': bool(RESEND_API_KEY)
    })

@app.route('/send-email', methods=['POST', 'OPTIONS'])
def send_email():
    if request.method == 'OPTIONS':
        return '', 200
    
    try:
        data = request.get_json()
        print(f"Received email request: {data}")
        
        # Validate input
        if not data or not data.get('to') or not data.get('subject') or not data.get('html'):
            return jsonify({'error': 'Missing required fields: to, subject, html'}), 400
        
        # Prepare request to Resend API
        resend_data = {
            'from': data.get('from', FROM_EMAIL),
            'to': data['to'] if isinstance(data['to'], list) else [data['to']],
            'subject': data['subject'],
            'html': data['html']
        }
        
        # Check if Resend is properly configured
        if not RESEND_API_KEY:
            return jsonify({
                'error': 'Resend API key not configured',
                'details': 'Please set RESEND_API_KEY in environment variables'
            }), 500
        
        # Send email using Resend API
        headers = {
            'Authorization': f'Bearer {RESEND_API_KEY}',
            'Content-Type': 'application/json'
        }
        
        response = requests.post(
            'https://api.resend.com/emails',
            json=resend_data,
            headers=headers
        )
        
        if response.status_code == 200:
            result = response.json()
            print(f"Email sent successfully: {result}")
            return jsonify({'success': True, 'data': result})
        else:
            error_text = response.text
            print(f"Email sending failed: {error_text}")
            return jsonify({'error': 'Failed to send email', 'details': error_text}), response.status_code
            
    except Exception as error:
        print(f"Email sending error: {error}")
        return jsonify({'error': 'Failed to send email', 'details': str(error)}), 500

if __name__ == '__main__':
    PORT = int(os.getenv('PORT', 3001))
    print(f"Email proxy server running on port {PORT}")
    print(f"Health check: http://localhost:{PORT}/health")
    print(f"Send email endpoint: http://localhost:{PORT}/send-email")
    
    # Check if API key is configured
    if not RESEND_API_KEY:
        print("WARNING: RESEND_API_KEY is not set in environment variables")
        print("Email sending will not work until you configure the API key")
    else:
        print("Resend API key is configured correctly")
    
    app.run(host='0.0.0.0', port=PORT, debug=False)